### PR TITLE
Add `(glob_files <glob>)` and `(glob_files_rec <glob>)` to the `files` field of the `install` stanza

### PR DESCRIPTION
**Pull request by gridbugs**
Closes https://github.com/ocaml/dune/issues/6018

### CODE REVIEW

File: src/dune_rules/glob_files.ml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+open! Import
+
+type t =
+  { glob : String_with_vars.t
+  ; recursive : bool
+  }
+
+module Path_split = struct
+  (* A source path, or a build path which has been split into its _build/foo
```
Comment (christinerose) [inline src/dune_rules/glob_files.ml:9]:
```suggestion
  (* A source path, or a build path, which has been split into its _build/foo
```
Either a comma after build path or no comma after source path.

File: src/dune_rules/glob_files.ml [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+     prefix, and its remainder. This allows us to generalize over paths in the
```
Comment (christinerose) [inline src/dune_rules/glob_files.ml:10]:
```suggestion
     prefix and its remainder. This allows us to generalize over paths in the
```
Does Dune files stick with the British spelling as we do with Tarides? If so, generalise instead of -ize.

File: test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t [hunk 3]
```diff
@@ -0,0 +1,33 @@
+Test that the glob_files terms are only accepted in the files field and not in
```
Comment (christinerose) [inline test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t:1]:
```suggestion
Test that the ``glob_files`` terms are only accepted in the ``files`` field and not in
```

File: test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+the dirs field
```
Comment (christinerose) [inline test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t:2]:
```suggestion
the ``dirs`` field
```

File: test/blackbox-tests/test-cases/install-glob/install-glob-recursive.t [hunk 5]
```diff
@@ -0,0 +1,69 @@
+Test that glob_files_rec can recursively find files
```
Comment (christinerose) [inline test/blackbox-tests/test-cases/install-glob/install-glob-recursive.t:1]:
```suggestion
Test that ``glob_files_rec`` can recursively find files
```

**Review by christinerose — CHANGES_REQUESTED** [review]
A few minor grammar / formatting suggestions as well as a question or two.

File: src/dune_rules/dep_conf.ml [hunk 6]
```diff
@@ -51,11 +48,12 @@ let decode =
       ; ("alias", sw >>| fun x -> Alias x)
       ; ("alias_rec", sw >>| fun x -> Alias_rec x)
       ; ( "glob_files"
-        , sw >>| fun x -> Glob_files { glob = x; recursive = false } )
+        , sw >>| fun glob -> Glob_files Glob_files.{ glob; recursive = false }
```
Comment (rgrinberg) [inline src/dune_rules/dep_conf.ml:51]: Should be `{ Glob_files.glob ; recursive = false }`

File: src/dune_rules/glob_files.mli [hunk 7]
```diff
@@ -0,0 +1,17 @@
+open! Import
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.mli:1]: Add a small description of the module please.

File: test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t (hunk 3 above)
Comment (rgrinberg) [reply test/blackbox-tests/test-cases/install-glob/install-glob-files-only.t:1]: This syntax is only available in .rst files

File: src/dune_rules/dune_file.ml [hunk 8]
```diff
@@ -1006,9 +1006,56 @@ end
...
+          in
+          let glob_loc = String_with_vars.loc glob_files.glob in
+          List.map paths ~f:(fun path ->
+              let src = (glob_loc, Path.to_string path) in
+              File_binding.Unexpanded.make ~src ~dst:src)
+
+      let to_file_bindings_expanded t ~expand_str ~dir =
+        let open Memo.O in
+        to_file_bindings_unexpanded t ~expand_str
+        >>= Memo.List.map ~f:(File_binding.Unexpanded.expand ~dir ~f:expand_str)
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.ml:1053]: It's a matter of style, but you can just do `Memo.bind` and remove the `let open` uses only once.

File: src/dune_rules/dep_conf.ml [hunk 9]
```diff
@@ -51,11 +48,12 @@ let decode =
...
+        )
       ; ( "glob_files_rec"
         , let+ () = Dune_lang.Syntax.since Stanza.syntax (3, 0)
-          and+ x = sw in
-          Glob_files { glob = x; recursive = true } )
+          and+ glob = sw in
+          Glob_files Glob_files.{ glob; recursive = true } )
```
Comment (rgrinberg) [inline src/dune_rules/dep_conf.ml:56]: ditto

**Comment by rgrinberg** [comment]
Should be updated to require 3.6.0

Btw, does this compose with `include`?

File: src/dune_rules/dep_conf.ml (hunk 6 above)
Comment (gridbugs) [reply src/dune_rules/dep_conf.ml:51]: Is including the module name explicitly here just to make the code more clear and so it doesn't look like the records fields are arguments to the `Glob_files` constructor?

Do we use the convention `{ Module_name.first_field; ... }` over `Module_name.{ first_field; ... }`?

Comment (rgrinberg) [reply src/dune_rules/dep_conf.ml:51]: > Do we use the convention { Module_name.first_field; ... } over Module_name.{ first_field; ... }?

They have slightly different meanings. The former is for qualifying labels when constructing/matching records, while the latter is a local open - equivalent to `let open Module_name in { first_field ; ... }`

Comment (gridbugs) [reply src/dune_rules/dep_conf.ml:51]: Right that makes sense. Is the former preferred because using a local open would do more than is strictly necessary, and it might take longer to compile or bring unwanted things into scope?

File: src/dune_rules/glob_files.ml (hunk 1 above)
Comment (gridbugs) [reply src/dune_rules/glob_files.ml:9]: I've reworded it slightly to clarify that the path is only split in the case of a build path.

File: src/dune_rules/glob_files.ml (hunk 2 above)
Comment (gridbugs) [reply src/dune_rules/glob_files.ml:10]: Updated to use the British spelling.

File: test/blackbox-tests/test-cases/install-glob/install-glob-recursive.t (hunk 5 above)
Comment (gridbugs) [reply test/blackbox-tests/test-cases/install-glob/install-glob-recursive.t:1]: For source code quoted in comments in .t files, I've switched to single backticks since that seems to be the convention.

**Comment by gridbugs** [comment]
I added a test which used globs in a dune file inside a subdirectory and that highlighted a problem with my implementation. I've updated this so that now the `Glob_files.expand_*` functions take an additional `base_dir` argument, and the globs are expanded relative to this argument. This involved reimplementing the `Glob_files` module. Apologies for the noise.

File: test/blackbox-tests/test-cases/glob_files_rec.t [hunk 10]
```diff
@@ -9,7 +9,7 @@ be release as it. We plan to replace it by recursive globs for 3.0.0.
   > (rule
   >  (alias x)
   >  (deps (glob_files_rec foo/*.txt))
-  >  (action (bash "for i in %{deps}; do printf \"\$i\\n\"; done")))
+  >  (action (bash "for i in %{deps}; do printf \"\$i\\n\"; done | sort")))
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/glob_files_rec.t:12]: I'm surprised this is necessary. If dune doesn't return `%{deps}` in some deterministic order, then it would certainly be a bug in dune.

File: src/dune_rules/glob_files.ml (hunk 1 above)
Comment (christinerose) [reply src/dune_rules/glob_files.ml]: Great!

**Review by christinerose — APPROVED** [review]

File: test/blackbox-tests/test-cases/glob_files_rec.t (hunk 10 above)
Comment (gridbugs) [reply test/blackbox-tests/test-cases/glob_files_rec.t]: Ah good point. I've updated `Glob_files.expand_*` to sort their results internally.

**Comment by gridbugs** [comment]
> Updated to require 3.6. Good question about `include`. It _should_, but I'll add a test tomorrow to confirm.

Added test composing `glob_files` and `include`

**Comment by rgrinberg** [comment]
The failure in CI looks a little suspicious:

```
#=== ERROR while compiling menhirLib.20220210 =================================#
# context     2.1.3 | linux/x86_64 | ocaml-base-compiler.4.14.0 | git+https://github.com/ocaml/opam-repository.git
# path        ~/work/dune/dune/_opam/.opam-switch/build/menhirLib.20220210
# command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p menhirLib -j 2
# exit-code   1
# env-file    ~/.opam/log/menhirLib-5720-7a711d.env
# output-file ~/.opam/log/menhirLib-5720-7a711d.out
### output ###
# Error: path outside the workspace: ../*.{ml,mli} from .
# -> required by _build/default/lib/pack/menhirLib.ml
# -> required by _build/install/default/lib/menhirLib/menhirLib.ml
# -> required by _build/default/menhirLib.install
# -> required by alias install
```

Do you think it's related to some your changes?

**Comment by gridbugs** [comment]
> The failure in CI looks a little suspicious:
> 
> ```
> #=== ERROR while compiling menhirLib.20220210 =================================#
> # context     2.1.3 | linux/x86_64 | ocaml-base-compiler.4.14.0 | git+https://github.com/ocaml/opam-repository.git
> # path        ~/work/dune/dune/_opam/.opam-switch/build/menhirLib.20220210
> # command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p menhirLib -j 2
> # exit-code   1
> # env-file    ~/.opam/log/menhirLib-5720-7a711d.env
> # output-file ~/.opam/log/menhirLib-5720-7a711d.out
> ### output ###
> # Error: path outside the workspace: ../*.{ml,mli} from .
> # -> required by _build/default/lib/pack/menhirLib.ml
> # -> required by _build/install/default/lib/menhirLib/menhirLib.ml
> # -> required by _build/default/menhirLib.install
> # -> required by alias install
> ```
> 
> Do you think it's related to some your changes?

Interesting. I'll investigate this tomorrow.

**Comment by gridbugs** [comment]
I found yet another case which this didn't handle. Namely, it would crash if the glob contained an absolute path. Note that relative globs containing absolute paths previously resulted in an unhandled error. I've added a `User_error` for this case instead. I still need to add a test for the new error, but I tried earlier today and the `loc` underlining was too wide. Need to investigate that further. I would have caught this case sooner (we do have a test for absolute globs) but when I run the test suite locally (tried on nixos and a very vanilla ubuntu docker image) many of the tests fail and I'm not sure why. Will investigate this tomorrow too.

**Comment by gridbugs** [comment]
I found two cases which would previously crash dune (prior to this PR that is):
- absolute paths to files or directories in the `install` stanza
- absolute paths in `(glob_files_rec ...)` (in `(rule (deps ...))`)

I've made these cases result in user errors instead and updated the docs accordingly.

**Comment by gridbugs** [comment]
> > The failure in CI looks a little suspicious:
> > ```
> > #=== ERROR while compiling menhirLib.20220210 =================================#
> > # context     2.1.3 | linux/x86_64 | ocaml-base-compiler.4.14.0 | git+https://github.com/ocaml/opam-repository.git
> > # path        ~/work/dune/dune/_opam/.opam-switch/build/menhirLib.20220210
> > # command     ~/.opam/opam-init/hooks/sandbox.sh build dune build -p menhirLib -j 2
> > # exit-code   1
> > # env-file    ~/.opam/log/menhirLib-5720-7a711d.env
> > # output-file ~/.opam/log/menhirLib-5720-7a711d.out
> > ### output ###
> > # Error: path outside the workspace: ../*.{ml,mli} from .
> > # -> required by _build/default/lib/pack/menhirLib.ml
> > # -> required by _build/install/default/lib/menhirLib/menhirLib.ml
> > # -> required by _build/default/menhirLib.install
> > # -> required by alias install
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Do you think it's related to some your changes?
> 
> Interesting. I'll investigate this tomorrow.

I think I fixed it. I was confused about what a `Path.t` really represents and had been using `Path.Source.t` to represent a relative path. This was incorrect as this type carries the assumption that it's a path relative to the project root, not relative to an arbitrary directory inside the project. Thus paths like `..` can't be represented since, interpreting them relative to the project root, they would take us outside the project. Globs need the ability to represent paths beginning with ".." to handle globs in subdirectories refering to files outside that subdirectory. I've changed it to use `string`s to represent relative paths.

**Comment by gridbugs** [comment]
Temporarily subsuming https://github.com/ocaml/dune/pull/6321 because this change now depends on that fix. Don't merge until after https://github.com/ocaml/dune/pull/6321 is merged.

**Comment by rgrinberg** [comment]
Why aren't absolute paths allowed in install stanzas? They're allowed in dependency specifications.

**Comment by gridbugs** [comment]
> Why aren't absolute paths allowed in install stanzas? They're allowed in dependency specifications.

Just because currently they cause dune to crash. `File_binding.Unexpanded.expand` assumes that its input is a relative path. I think this is a bug and that they should be supported but I didn't want to fix that bug as part of this work.

File: src/dune_rules/dune_file.ml [hunk 11]
```diff
@@ -1076,10 +1076,74 @@ module Plugin = struct
...
+          let+ file_binding = File_binding.Unexpanded.decode in
+          File_binding file_binding
+        in
+        let glob_files_decode =
+          let version_check = Dune_lang.Syntax.since Stanza.syntax (3, 6) in
+          let+ glob_files =
+            sum
+              [ ( "glob_files"
+                , version_check >>> String_with_vars.decode >>| fun glob ->
+                  Glob_files.{ glob; recursive = false } )
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.ml:1107]: `{ Glob_files.glob ; recursive = false } ` to avoid the local open

File: src/dune_rules/dune_file.ml [hunk 12]
```diff
@@ -1076,10 +1076,74 @@ module Plugin = struct
...
+                , version_check >>> String_with_vars.decode >>| fun glob ->
+                  Glob_files.{ glob; recursive = true } )
+              ]
+          in
+          Glob_files glob_files
+        in
+        file_binding_decode <|> glob_files_decode
+
+      let to_file_bindings_unexpanded t ~expand_str ~dir =
+        let open Memo.O in
```
Comment (rgrinberg) [inline src/dune_rules/dune_file.ml:1118]: this `let open` can be moved to the `Glob_files` clause.

File: src/dune_rules/glob_files.ml [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+     string manipulation rather than [Filename.dirname] so that the result can
+     be used to reconstruct strings representing paths to files matched by the
+     glob in the style of the original glob. For example, the globs "*" and
+     "./*" will match the same files, but we want the results of the latter to
+     include the "./" prefix, but not the former. *)
+  let parent_str =
+    match String.drop_suffix glob_string ~suffix:pattern_str with
+    | Some x -> x
+    | None ->
+      failwith
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.ml:62]: We use `Code_error.raise` instead of `failwith` for programming errors.

File: src/dune_rules/glob_files.ml [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+
+let expand_action_builder t ~f ~base_dir =
+  let open Action_builder.O in
+  let loc = String_with_vars.loc t.glob in
+  let* without_vars = Expand_vars_action_builder.expand_vars t ~f ~base_dir in
+  Without_vars.file_selectors_with_relative_dirs without_vars ~loc
+  |> Action_builder.of_memo
+  >>= Action_builder.List.concat_map ~f:(fun (file_selector, relative_dir) ->
+          Action_builder.paths_matching ~loc file_selector
+          >>| Path.Set.to_list
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.ml:153]: `to_list_map` for a minor optimization.

File: src/dune_rules/glob_files.ml [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+  >>| List.sort ~compare:String.compare
+
+let expand_memo t ~f ~base_dir =
+  let open Memo.O in
+  let* without_vars = Expand_vars_memo.expand_vars t ~f ~base_dir in
+  let loc = String_with_vars.loc t.glob in
+  Without_vars.file_selectors_with_relative_dirs without_vars ~loc
+  >>= Memo.List.concat_map ~f:(fun (file_selector, relative_dir) ->
+          Build_system.eval_pred file_selector
+          >>| Path.Set.to_list
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.ml:164]: ditto

File: src/dune_rules/glob_files.ml [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+  let* without_vars = Expand_vars_action_builder.expand_vars t ~f ~base_dir in
+  Without_vars.file_selectors_with_relative_dirs without_vars ~loc
+  |> Action_builder.of_memo
+  >>= Action_builder.List.concat_map ~f:(fun (file_selector, relative_dir) ->
+          Action_builder.paths_matching ~loc file_selector
+          >>| Path.Set.to_list
+          >>| List.map ~f:(replace_path_dir relative_dir))
+  >>| List.sort ~compare:String.compare
+
+let expand_memo t ~f ~base_dir =
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.ml:157]: This function looks awfully similar, can we only keep one if possible?

File: src/dune_rules/glob_files.mli [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+(* A glob stored in a [String_with_vars.t] and functions for expanding the glob
```
Comment (rgrinberg) [inline src/dune_rules/glob_files.mli:3]: The comments should be ocamldoc comments. That is, `(** .. *)`.

File: test/blackbox-tests/test-cases/glob-deps-outside-directory.t [hunk 18]
```diff
@@ -0,0 +1,20 @@
+Glob deps in a subdirectory that refer to a path outside that subdirectory
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/glob-deps-outside-directory.t:1]: Can this be moved to a separate PR?

File: test/blackbox-tests/test-cases/glob_files_rec.t [hunk 19]
```diff
@@ -82,3 +82,26 @@ Check that generated directories are ignored
 
   $ dune build @x
 
+Check that we get a user error if an absolute path is passed to glob_files_rec
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/glob_files_rec.t:85]: Separate PR if possible please.

File: test/blackbox-tests/test-cases/glob_files_rec.t [hunk 20]
```diff
@@ -82,3 +82,26 @@ Check that generated directories are ignored
...
+------------------------------------------------------------------------------
+
+Store the current directory in a file so the underline in the error message is of
+a consistent length across different systems.
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/glob_files_rec.t:89]: Nice trick!

File: test/blackbox-tests/test-cases/glob_files_rec.t (hunk 19 above)
Comment (gridbugs) [reply test/blackbox-tests/test-cases/glob_files_rec.t]: moved to https://github.com/ocaml/dune/pull/6331

File: test/blackbox-tests/test-cases/glob-deps-outside-directory.t (hunk 18 above)
Comment (gridbugs) [reply test/blackbox-tests/test-cases/glob-deps-outside-directory.t:1]: moved to https://github.com/ocaml/dune/pull/6332

File: src/dune_rules/glob_files.ml (hunk 16 above)
Comment (gridbugs) [reply src/dune_rules/glob_files.ml]: I've had a go at combining them with a functor, and added a `Glob_file.Expand` signature to avoid repetition in the interface. Can you take a look and tell me if there's too much indirection for it to be readable? It's non-trivial because it needs to generalize over the `Memo.t` and `Action_builder.t` type constructors. Is there are more ergonomic way of doing this than what I've come up with? I tried to simplify it with first-class modules but ran into an issue because I couldn't access internal parameterized types of first-class modules in function signatures.

**Comment by gridbugs** [comment]
> It would also be good to separate all the absolute paths business into a separate PR.

Agreed. I've opened https://github.com/ocaml/dune/pull/6331 and rebased this PR on top.

**Comment by rgrinberg** [comment]
Merged after a few cosmetic changes and simplifications. Thanks.
